cryptomator/src/main/java/ch/cyberduck/core/cryptomator/impl/CryptoDirectoryUVFProvider.java

```diff
@@ -0,0 +1,156 @@
+import ch.cyberduck.core.cryptomator.CryptorCache;
+import ch.cyberduck.core.exception.BackgroundException;
+import ch.cyberduck.core.exception.NotfoundException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.cryptomator.cryptolib.api.FileHeader;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.EnumSet;
+
+public class CryptoDirectoryUVFProvider extends CryptoDirectoryV7Provider {
+    private static final Logger log = LogManager.getLogger(CryptoDirectoryUVFProvider.class);
+
+    private final Path home;
+    private final AbstractVault vault;
+
+    private final RandomStringService random
+            = new UUIDRandomStringService();
+    private final Path dataRoot;
+    private final CryptorCache filenameCryptor;
+
+    public CryptoDirectoryUVFProvider(final AbstractVault vault, final CryptoFilename filenameProvider, final CryptorCache filenameCryptor) {
+        super(vault, filenameProvider, filenameCryptor);
+        this.filenameCryptor = filenameCryptor;
+        this.home = vault.getHome();
+        this.vault = vault;
+        this.dataRoot = new Path(home, "d", home.getType());
+    }
+
+    @Override
+    protected byte[] toDirectoryId(final Session<?> session, final Path directory, final byte[] directoryId) throws BackgroundException {
+        if(new SimplePathPredicate(home).test(directory)) {
+            return vault.getRootDirId();
+        }
+        return super.toDirectoryId(session, directory, directoryId);
+    }
+
+    @Override
+    public Path toEncrypted(final Session<?> session, final byte[] directoryId, final Path directory) throws BackgroundException {
+        if(!directory.isDirectory()) {
+            throw new NotfoundException(directory.getAbsolute());
+        }
+        if(new SimplePathPredicate(directory).test(home) || directory.isChild(home)) {
+            final PathAttributes attributes = new PathAttributes(directory.attributes());
+            // The root of the vault is a different target directory and file ids always correspond to the metadata file
+            attributes.withVersionId(null);
+            attributes.withFileId(null);
+            // Remember random directory id for use in vault
+            final byte[] id = this.toDirectoryId(session, directory, directoryId);
+            log.debug("Use directory ID '{}' for folder {}", id, directory);
+            attributes.setDirectoryId(id);
+            attributes.setDecrypted(directory);
+            final String directoryIdHash;
+            if(new SimplePathPredicate(home).test(directory)) {
+                // TODO hard-coded to initial seed in UVFVault
+                directoryIdHash = filenameCryptor.hashDirectoryId(id);
+            }
+            else {
+                directoryIdHash = vault.getCryptor().fileNameCryptor(loadRevision(session, directory)).hashDirectoryId(id);
+            }
```

Comment:
@ylangisc what's the best approach - should we cache the revisions per directory along with its ID?